src/inline-assembly.md

```diff
@@ -37,515 +52,937 @@ assert_eq!(x, 4 * 6);
+
+r[asm.invocation.duplication]
+The number of times, locations, and the order in which a given invocation of [`core::arch::asm!`] is expanded is unspecified.
+
+```rust,ignore
+// The following code may have suprising results, and may fail to compile or link. 
+// The results, including whether it succesfully compiles, may depend on non-local use sites of the function, and on optimization settings.
+# #[cfg(target_arch = "x86_64")] { unsafe{
+let x: i32;
+core::arch::asm!("foo: jmp foo");
+# }}
+```
+
+> [!NOTE]
+> In particular, an asm block may be duplicated, for example if the containing function is inlined, or omitted from the output entirely.
+> As a consequence, asm blocks should not use directives that have non-idempotent non-local effects, or named labels and symbol definitions. 
+> Additionally, two asm blocks may not rely upon being adjacent in executable memory, even if they are adjacent in the source.
+
+> [!NOTE]
+> Local Labels (a decimal literal) may be used freely if the asm block needs to define a label. Due to a bug, literals that solely consist of 1s and 0s are not valid local labels.
+> See [The GNU AS Manual on Local Labels](https://sourceware.org/binutils/docs/as/Symbol-Names.html) for details on local labels.
+> It is not guaranteed that a local label defined in one asm block will be accessible from an adjacent asm block.
+
+```rust,no_run
+# #[cfg(target_arch = "x86_64")] { unsafe{
+let x: i32;
+core::arch::asm!("2: jmp 2b");
+# }}
+```
+
+r[asm.invocation.global-order]
+The order in which invocations of [`core::arch::global_asm!`] are expanded is unspecified.
+
+r[asm.invocation.directive-state]
+The *expanded asm-string* shall not issue a directive that modifies the global state of the assembler for processing inputs unless it issues a directive to restore that state it had upon entering the block. No diagnostic is required.
+
+> [!NOTE]
+> This include state such as the current section of the assembler, the syntax mode, or the kind of assembly output being generated.
+> Failing to obey this requirement can have significant impact on code generation, including code unrelated to the asm block. For example, an asm block that issues a `.data` directive without resetting to the appropriate section for the function can cause the following code in the function to be generated in the `.data` section, and for execution to fall off the asm block into improper memory.
+
+r[asm.invocation.global-section]
+The *expanded asm-string* of a [`core::arch::global_asm!`] invocation acts as though a `.section` directive is issued before the *expanded asm-string*  which causes code to be generated in the default section on the target for executable code.
+
+> [!NOTE]
+> This section is typically named `.text`. 
+
+
+r[asm.invocation.prefix-instr]
+An *expanded asm-string* shall not end with an instruction that is interpreted as a prefix on the architecture. No Diagnostic is required
```

Comment:
```suggestion
An *expanded asm-string* shall not end with an instruction that is interpreted as a prefix on the architecture. No diagnostic is required.
```

Also, I'm not sure why it is mentioning whether a diagnostic is required. Can you say more why it says that?

Reply:
The presumption is that a diagnostic is required, and that the compiler is forbidden to compile an ill-formed program (one that violates a constraint like that one). That overrides the rule by telling the compiler that it's free to ignore the violation and translate the program notwithstanding, but that the result of the program is allowed to be anything.